Framework/Geometry/src/Instrument/DetectorInfo.cpp

```diff
@@ -228,6 +228,52 @@ DetectorInfo::signedTwoTheta(const std::pair<size_t, size_t> &index) const {
   return angle;
 }
 
+double DetectorInfo::azimuthal(const size_t index) const {
+  if (isMonitor(index))
+    throw std::logic_error("Azimuthal angle is not defined for monitors");
+
+  const auto samplePos = samplePosition();
+  const auto beamLine = samplePos - sourcePosition();
+
+  if (beamLine.nullVector()) {
+    throw Kernel::Exception::InstrumentDefinitionError(
+        "Source and sample are at same position!");
+  }
+
+  const auto sampleDetVec = position(index) - samplePos;
+  const auto horizontal =
+      m_instrument->getReferenceFrame()->vecPointingHorizontal();
+
+  const double dotHorizontal = sampleDetVec.scalar_prod(horizontal);
+  const auto vertical = Kernel::normalize(beamLine).cross_prod(horizontal);
```

Comment:
Depending on the case, won't this point the direction *opposite* to the original "vertical" defined in the reference frame? See how https://github.com/mantidproject/mantid/blob/master/Framework/Geometry/src/Instrument/ReferenceFrame.cpp#L133 is implemented.